### PR TITLE
Document cidrlist parameter deprecation

### PR DESCRIPTION
**Pull request by onitake**
## Description

Amend the description for the cidrlist parameter in CreatePortForwardingRule and CreateFirewallRule to state that it is deprecated and shouldn't be used.

Passing this parameter will always result in an error.

Fixes: #4036 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

See the original bug report for an explanation.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@onitake can you update this (and maybe make sure the cidr lists are actually ignored and don't throw an exception)?

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@onitake conflicts on this PR, please fix.

**Comment by rhtyd** [comment]
ping @onitake is this still relevant?

**Comment by sureshanaparti** [comment]
Hi @onitake Can you please resolve conflicts in this PR?

**Comment by onitake** [comment]
I combined the documentation strings. Hope that's ok?

**Review by DaanHoogland — APPROVED** [review]
doc only; lgtm

**Comment by nvazquez** [comment]
@onitake looks good, only one question: if the parameter is deprecated and must not be used why not then remove it? cc. @rhtyd @DaanHoogland

**Comment by DaanHoogland** [comment]
> @onitake looks good, only one question: if the parameter is deprecated and must not be used why not then remove it? cc. @rhtyd @DaanHoogland

For backward compatibility reasons for scripts, but you are right @nvazquez. Unknown parameters are ignored, so if we are sure it is completely ignored. I didn't think of that. I'd say let's keep them in like this for a few versions and then remove them.

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreateFirewallRuleCmd.java [hunk 1]
```diff
@@ -73,7 +73,7 @@
     @Parameter(name = ApiConstants.END_PORT, type = CommandType.INTEGER, description = "the ending port of firewall rule")
     private Integer publicEndPort;
 
-    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the CIDR list to forward traffic from. Multiple entries must be separated by a single comma character (,).")
+    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the CIDR list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is depcreated. Do not use.")
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreateFirewallRuleCmd.java:76]: typo, _depcreated_ => _deprecated_

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java [hunk 2]
```diff
@@ -108,7 +108,7 @@
                 description = "the ID of the virtual machine for the port forwarding rule")
     private Long virtualMachineId;
 
-    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the cidr list to forward traffic from. Multiple entries must be separated by a single comma character (,).")
+    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the cidr list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is depcreated. Do not use.")
```
Comment (sureshanaparti) [inline api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:111]: same typo error here as well

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreateFirewallRuleCmd.java (hunk 1 above)
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreateFirewallRuleCmd.java:76]:
```suggestion
    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the CIDR list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is deprecated. Do not use.")
```

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java (hunk 2 above)
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:111]:
```suggestion
    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the cidr list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is deprecated. Do not use.")
```

**Comment by nvazquez** [comment]
Ok thanks @DaanHoogland 
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by onitake** [comment]
@sureshanaparti @DaanHoogland Woops, thank you. :sweat_smile:

**Review by sureshanaparti — APPROVED** [review]
changes LGTM

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
Merging this based on approvals and test results

**Comment by weizhouapache** [comment]
@sureshanaparti @DaanHoogland @onitake 
I think cidrlist is not deprecated in CreateFirewallRuleCmd.java. 
Please revert this merge.

@onitake could you please create another pr for CreatePortForwardingRuleCmd.java which cidrlist is deprecated ?

```
(localcloud) SBCM5> > create firewallrule ipaddressid=997f8efd-2ec7-46f4-a054-de837e196cc2 startport=22 endport=22 cidrlist=10.10.10.0/24 protocol=tcp
{
  "firewallrule": {
    "cidrlist": "10.10.10.0/24",
    "endport": 22,
    "fordisplay": true,
    "id": "f0b7270d-77b0-4d97-8c02-d95a0e774561",
    "ipaddress": "10.0.52.5",
    "ipaddressid": "997f8efd-2ec7-46f4-a054-de837e196cc2",
    "networkid": "5528f815-c365-4702-8a6b-b59a1d33e8c5",
    "protocol": "tcp",
    "startport": 22,
    "state": "Active",
    "tags": []
  }
}
```

**Comment by onitake** [comment]
@weizhouapache Couldn't you have said this a few days earlier? This PR has been open for more than a year. :disappointed: 

Sure, I can submit another PR that reverts this... But only if it's confirmed that this isn't deprecated. @rhtyd @DaanHoogland ?

**Comment by weizhouapache** [comment]
> @weizhouapache Couldn't you have said this a few days earlier? This PR has been open for more than a year. 
> 
> Sure, I can submit another PR that reverts this... But only if it's confirmed that this isn't deprecated. @rhtyd @DaanHoogland ?

@onitake I found the problem right after this pr is merged...

maybe it is unnecessary to revert the pr as it is not an major issue, you can create a PR to (1) revert the change in CreateFirewallRuleCmd.java; (2) change CreateLoadBalancerRuleCmd.java and CreateIpForwardingRuleCmd.java

**Comment by weizhouapache** [comment]
@onitake 
as I mentioned above, could you please create another pr to fix the issue ? 
it is not urgent, but do not forget it.

**Pull request merged**